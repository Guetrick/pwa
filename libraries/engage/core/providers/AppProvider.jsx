import React, { Component } from 'react';
import PropTypes from 'prop-types';
import appConfig from '@shopgate/pwa-common/helpers/config';
import {
  APP_EVENT_VIEW_DID_APPEAR,
  APP_EVENT_VIEW_DID_DISAPPEAR,
} from '@shopgate/pwa-core/constants/AppEvents';
import { event } from '../index';
import AppContext from '../contexts/AppContext';

/**
 * The AppProvider component.
 */
class AppProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  }

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      isForeground: true,
    };

    event.addCallback(APP_EVENT_VIEW_DID_APPEAR, this.setForeground);
    event.addCallback(APP_EVENT_VIEW_DID_DISAPPEAR, this.setBackground);
  }

  /**
   * Removes the event listeners when the component unmounts.
   */
  componentWillUnmount() {
    event.removeCallback(APP_EVENT_VIEW_DID_APPEAR, this.setForeground);
    event.removeCallback(APP_EVENT_VIEW_DID_DISAPPEAR, this.setBackground);
  }

  /**
   * Returns the context value to be passed to consumers.
   * @returns {Object}
   */
  get provided() {
    return {
      appConfig,
      isForeground: this.state.isForeground,
      setIsForeground: this.setIsForeground,
    };
  }

  /**
   * Sets the isForeground state to TRUE.
   */
  setForeground = () => {
    this.setIsForeground(true);
  }

  /**
   * Sets the isForeground state to TRUE.
   */
  setBackground = () => {
    this.setIsForeground(false);
  }

  /**
   * Sets the isForeground state.
   * @param {boolean} value The new value
   */
  setIsForeground(value) {
    this.setState({
      isForeground: value,
    });
  }

  /**
   * @returns {JSX}
   */
  render() {
    return (
      <AppContext.Provider value={this.provided}>
        {this.props.children}
      </AppContext.Provider>
    );
  }
}

export default AppProvider;
