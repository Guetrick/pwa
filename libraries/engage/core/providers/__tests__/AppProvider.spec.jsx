import React from 'react';
import { shallow } from 'enzyme';
import appConfig from '@shopgate/pwa-common/helpers/config';
import event from '@shopgate/pwa-core/classes/Event';
import {
  APP_EVENT_VIEW_WILL_APPEAR,
  APP_EVENT_VIEW_WILL_DISAPPEAR,
} from '@shopgate/pwa-core/constants/AppEvents';
import AppProvider from '../AppProvider';

jest.mock('@shopgate/pwa-common/helpers/config', () => ({
  some: 'prop',
}));

jest.mock('@shopgate/pwa-core/classes/Event', () => ({
  addCallback: jest.fn(),
  removeCallback: jest.fn(),
}));

// eslint-disable-next-line require-jsdoc
const MockComponent = () => null;

describe('engage > core > providers > AppProvider', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should initialize as expected', () => {
    const wrapper = shallow((
      <AppProvider>
        <MockComponent />
      </AppProvider>
    ));

    const instance = wrapper.instance();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).length).toEqual(1);
    expect(wrapper.prop('value')).toEqual({
      appConfig,
      isForeground: true,
      setIsForeground: instance.setIsForeground,
    });
    expect(event.addCallback).toHaveBeenCalledTimes(2);
    expect(event.addCallback).toHaveBeenCalledWith(
      APP_EVENT_VIEW_WILL_APPEAR,
      instance.setForeground
    );
    expect(event.addCallback).toHaveBeenCalledWith(
      APP_EVENT_VIEW_WILL_DISAPPEAR,
      instance.setBackground
    );
  });

  it('should handle calls of setIsForeground() as expected', () => {
    const wrapper = shallow((
      <AppProvider>
        <MockComponent />
      </AppProvider>
    ));

    const instance = wrapper.instance();
    let { isForeground } = wrapper.prop('value');
    expect(isForeground).toBe(true);

    instance.setIsForeground(false);
    ({ isForeground } = wrapper.prop('value'));
    expect(isForeground).toBe(false);

    instance.setIsForeground(true);
    ({ isForeground } = wrapper.prop('value'));
    expect(isForeground).toBe(true);
  });

  it('should handle event calls as expected', () => {
    const wrapper = shallow((
      <AppProvider>
        <MockComponent />
      </AppProvider>
    ));

    const [[, setForeground], [, setBackground]] = event.addCallback.mock.calls;
    let { isForeground } = wrapper.prop('value');
    expect(isForeground).toBe(true);

    setBackground();
    ({ isForeground } = wrapper.prop('value'));
    expect(isForeground).toBe(false);

    setForeground();
    ({ isForeground } = wrapper.prop('value'));
    expect(isForeground).toBe(true);
  });

  it('should remove the event listeners on unmount', () => {
    const wrapper = shallow((
      <AppProvider>
        <MockComponent />
      </AppProvider>
    ));

    const instance = wrapper.instance();
    wrapper.unmount();

    expect(event.removeCallback).toHaveBeenCalledTimes(2);
    expect(event.removeCallback).toHaveBeenCalledWith(
      APP_EVENT_VIEW_WILL_APPEAR,
      instance.setForeground
    );
    expect(event.removeCallback).toHaveBeenCalledWith(
      APP_EVENT_VIEW_WILL_DISAPPEAR,
      instance.setBackground
    );
  });
});
