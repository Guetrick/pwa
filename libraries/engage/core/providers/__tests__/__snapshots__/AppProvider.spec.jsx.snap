// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > core > providers > AppProvider should initialize as expected 1`] = `
<ContextProvider
  value={
    Object {
      "appConfig": Object {
        "some": "prop",
      },
      "isVisible": true,
      "setIsVisible": [Function],
    }
  }
>
  <MockComponent />
</ContextProvider>
`;
